Infer needed `zero_alloc` checks from signatures

ccasin opened the pull request:
In the "`zero_alloc` in signatures" work, we made it so that to expose a function as zero alloc in your signature, you must also put a `zero_alloc` attribute on it in the structure.  This guarantees the backend check will run on that function, as implied by the signature.  But also it's a real pain to write the attribute in two places, and we can do better: This PR lets the user skip writing the attribute on the implementation and infers that the check is needed from the signature.

There are three commits, the first two of which are refactoring.  I suggest the main reviewer reads commit by commit.
1) The first commit makes a change that was suggested for the original signatures work.  It changes the way we store `zero_alloc` attributes on applications to be the syntactic version of those attributes rather than their values in the semantic domain.  At the same time, it changes when we check whether we can infer `[@zero_alloc assume]` on an application - now we do this during translation to lambda rather than when typechecking the application.  This is helpful in later commits - we are delaying that inference step until it is safe to default `zero_alloc` variables.
2) The second commit moves the logic for checking whether one `zero_alloc` attribute is at least as strong as another out of `Include_core` and into its own file.  It makes sense to have this together with the implementation of `zero_alloc` variables (which comes in the next commit), and I've made it its own commit so that I'm not both moving and changing that code in one commit.
3) The third commit implements inference and adds tests. If a function definition has no `zero_alloc` attribute on it we make a variable representing whether or not it will need to be checked.  That variable can be filled in when we compare the module to its signature.  These variables are somewhat simpler than jkind or type variables because we never need to link two of them.

I suggest @ncik-roberts reviews, as he did for the signatures implementation.  @gretay-js should take a look at the tests, at least.

This PR does not yet make it so that zero_alloc information is carried across aliases (`let f = g`) - there is enough here that I think that should be its own PR.

One minor design note:

### Locations in errors

Today `zero_alloc` errors look like this:

```
File "/home/ccasinghino/tmp/foo.ml", line 1, characters 5-15:
1 | let[@zero_alloc] f x = (x,x)
         ^^^^^^^^^^
Error: Annotation check for zero_alloc failed on function Foo.f (camlFoo__f_0_1_code)

File "/home/ccasinghino/tmp/foo.ml", line 1, characters 23-28:
1 | let[@zero_alloc] f x = (x,x)
                           ^^^^^
Error: allocation of 24 bytes
```

The error has two parts: The first identifies the location of the attribute that said to check something.  The second shows the offending allocation.

After this PR, the attribute that caused the check may be far removed from the implementation, off in a signature somewhere.  Making that show up first in the error seems bad for editor integration - people probably want to navigate to the offending function.  So if the check is an inferred one, I've made the first error point to the function that was checked instead.

I'm not sure that's exactly the right place to point.  Also it would probably be convenient to know what attribute caused this check, somehow, even if that's not going to be the place we most commonly want to navigate.  That information is easy to make available, but I think the right thing is to pass it to the backend and have the backend include it later in the error, and propose putting that off for a future PR.

ncik-roberts commented on tests/backend/zero_alloc_checker/test_inference.ml:
```diff
@@ -0,0 +1,73 @@
+let f_arity_two x = fun y -> x + y
+
+
+(* Shadowing: the mli marks [f_shadow] zero_alloc.  That check should only apply
+   to the last such function in the file, so we get no error even though there
+   is an earlier function with the same name and type that allocates. *)
```
Can you please write a test in the other direction, too? (You get an error if the shadowing function allocates, and the mli is marked zero-alloc?)

ncik-roberts commented on tests/backend/zero_alloc_checker/test_inference.ml:
```diff
@@ -0,0 +1,73 @@
+    let s = Printf.sprintf "%d\n" x in
+    raise (E s)
+
+(* These are tagged [@zero_alloc opt] in the mli - the latter should fail just
+   when the flag to check opt things is passed. *)
+let f_opt_success x =
+  if x = 42 then () else
+    let s = Printf.sprintf "%d\n" x in
+    raise (E s)
```
Is it easy to write a test of a function that is marked as `[@zero_alloc opt]` in the mli, and that _would_ fail if it was marked as `[@zero_alloc]`? (I.e. a test where the higher level of optimization removes allocation.)

ccasin replied:
I think no: For users we describe `opt` things as being checked when compiling with optimization, but actually it's just about whether you pass a flag saying the opt things should be checked (which people tell the build system to do when building with those higher optimization levels).

ncik-roberts reviewed (commented):
Just a review on tests so far.

ccasin commented:
I've reworked this slightly to offer nicer inference behavior, and rebased/squashed to give it a nice history.  Nothing about the original PR description needs changing.

One interesting note, pointed out by Leo, is that the rules for zero_alloc variables are similar to the rules for modalities on the comonadic (prescriptive) axes.  In typemod I've slightly generalized the infrastructure to support those that was just added in #2685 and renamed the functions to, hopefully, explain their purpose more clearly, make clear that they are doing something similar to `remove_mode_and_kind_variables`, and be clear they deal with both modalities and zero_alloc variables.

ncik-roberts commented on ocaml/testsuite/tests/typing-zero-alloc/signatures.ml:
```diff
@@ -1002,3 +936,257 @@ module type S_no_assume = sig val f : int -> int * int [@@zero_alloc] end
+  let f x = x+1
+end
+
+(* The current state of the variales is "no check" *)
```
```suggestion
(* The current state of the variables is "no check" *)
```

ncik-roberts commented on tests/backend/zero_alloc_checker/test_inference.mli:
```diff
@@ -0,0 +1,25 @@
+val[@zero_alloc opt] f_opt_fail : int -> int * int
+
+val[@zero_alloc strict opt] f_strict_opt_success : int -> int
+val[@zero_alloc strict opt] f_strict_opt_fail : int -> unit
```
I think we should add a test demonstrating the "spooky action at a distance" of a module inclusion check causing the zero-alloc check to flag an issue with the original module definition. (You have type-checking tests for something like this.)

e.g.
```ocaml
module type S = functor () -> sig val[@zero_alloc] f : 'a -> 'a option end

module F () = struct
  let f x = Some x
end

module _ : S = F
```

In particular, the error message does not identify why the zero-alloc check is running at all, and I think it would be good to have a test demonstrating the iffy error message (and even better to someday make it better).

ncik-roberts commented on ocaml/testsuite/tests/typing-zero-alloc/cmi_test.ml:
```diff
@@ -0,0 +1,77 @@
+       File "cmi_test_lib.ml", line 4, characters 4-28: Actual declaration
+|}]
+
+module M2 : sig
+  val[@zero_alloc strict] f : int -> int
+end = Cmi_test_lib.M_constrained_variable
+[%%expect{|
+Line 3, characters 6-41:
+3 | end = Cmi_test_lib.M_constrained_variable
+          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+Error: Signature mismatch:
+       Modules do not match:
+         sig val f : int -> int [@@zero_alloc] end
+       is not included in
+         sig val f : int -> int [@@zero_alloc strict] end
+       Values do not match:
+         val f : int -> int [@@zero_alloc]
+       is not included in
+         val f : int -> int [@@zero_alloc strict]
+       The former provides a weaker "zero_alloc" guarantee than the latter.
+       File "cmi_test_lib.ml", line 7, characters 6-7: Actual declaration
+|}]
```
Consider adding a test showing that `M_constrained_variable` is contained in `val[@zero_alloc] f : int -> int`.

ncik-roberts commented on ocaml/typing/typedtree.mli:
```diff
@@ -296,10 +296,10 @@ and expression_desc =
-            that attribute is absent, it is [Assume_info.none].
-          *)
+            The [zero_alloc_attribute] records the optional [@zero_alloc assume]
+            attribute that may appear on applications.  Invariant: it is either
+            [Default_zero_alloc] (if the user wrote no attribute) or [Assume]
+            (if they did). *)
```
Why not mint a new type and remove the failure cases? (I'm currently confused how `Ignore_assert_all` is ruled out by type-checking, and I think minting a new type here makes that easier to see/understand.)

ccasin replied:
I've taken a stab at this. It didn't result in any major simplfications, but is probably a bit clearer - let me know what you think.

ncik-roberts reviewed (commented):
some review

ncik-roberts commented on ocaml/typing/zero_alloc.mli:
```diff
@@ -0,0 +1,50 @@
+   done. *)
+val default : t
+
+val create : const -> t
```
Suggestion: `create_const`.

ncik-roberts commented on ocaml/typing/typedtree.ml:
```diff
@@ -1068,32 +1068,34 @@ let let_bound_idents_with_modes_sorts_and_checks bindings =
-                       loc = Location.none;
-                       opt = false }
+             if !Clflags.zero_alloc_check_assert_all && arity > 0 then
+               Zero_alloc.create (Check { strict = false;
+                                          arity;
+                                          loc = Location.none;
+                                          opt = false })
```
If you rename `Zero_alloc.create` to `Zero_alloc.create_const` as I suggested elsewhere, you'll be prompted to change the indentation of the labels of `Check`. Let's just move the constructor to a new line to reduce this kind of sensitivity.

ncik-roberts commented on ocaml/typing/typedtree.ml:
```diff
@@ -1068,32 +1068,34 @@ let let_bound_idents_with_modes_sorts_and_checks bindings =
+           | Ignore_assert_all | Check _ | Assume _ -> fn.zero_alloc
          in
          Ident.Map.add id zero_alloc checks
+         (* CR ccasinghino: we could copy the zero-allocness if the vb_expr
+            is an ident. *)
```
Flagging this: not sure if you intended on addressing this before merging.

ccasin replied:
Sorry - this was a note to myself for the next PR.  I've made the comment clearer.

ncik-roberts reviewed (commented):
Thanks for taking the time to add zero_alloc variables. The behavior seems much more normal to me know. I'll take a quick look again after you respond to my (mostly) nits.

ccasin commented on ocaml/typing/includecore.ml:
```diff
@@ -167,7 +98,9 @@ let value_descriptions ~loc env name
     vd1.val_attributes vd2.val_attributes
     name;
-  zero_alloc vd1.val_zero_alloc vd2.val_zero_alloc;
+  begin try Zero_alloc.sub_exn vd1.val_zero_alloc vd2.val_zero_alloc with
```
I think this is fairly idiomatic for the compiler (e.g., this isn't the only try/with in this function). `sub_exn` is a series of checks any of which may fail independently, and I wanted to avoid writing a bunch of gross nested matches.  I could define a letop for bind in the result monad, or something, but in the typechecker the usual style is just to raise.

(I don't feel particularly strongly though - feel free to push back more)

ncik-roberts replied:
I thought all error creation appeared in tail position, except for one `if`-`then` needing to grow an `else`. If that’s true, let’s change it to return an error.

ccasin replied:
Probably I'm missing some nice way to refactor this, but it's [these](https://github.com/ccasin/flambda-backend/blob/ccd41d34bd07feff3947fbaf2f96af0756789aac/ocaml/typing/zero_alloc.ml#L112) [two](https://github.com/ccasin/flambda-backend/blob/ccd41d34bd07feff3947fbaf2f96af0756789aac/ocaml/typing/zero_alloc.ml#L118) raises that seemed annoying.

ncik-roberts replied:
I missed the second one. Fine, then we should move the try-with into the module.

ccasin replied:
Good idea, done.

ccasin commented:
Thanks for the review.  I think I've addressed all the minor stuff and will play around with the types in typedtree sometime in the next few days.

ccasin commented:
> Thanks for the changes. I think the type refactoring you did in the latest commit is nice and helps make some non-obvious (or comment-enforced) invariants much clearer. I haven't checked whether additional type refactors would be nice in the same way, but you can make the call.

Thanks.  I did look for additional opportunities to make this kind of change, but didn't find them, because the other places where zero_alloc stuff appears in types or the typedtree need the option to be variables.
